### PR TITLE
🎁 Repeatable service to re-run failed entries

### PR DESCRIPTION
**Pull request by jeremyf**
During the normal lifecycle of an import, we inevitably encounter
errors.  Those errors can be from a variety of problems.

This service class, which is not part of production, is the beginnings
of the logic necessary for re-running only the failed entries while also
emulating having re-run all of the successful entries.

Part of the reason for the emulation is to ensure that counts, which are
already suspect, are consistent from run to run.

Note: I have prefaced this PR with the 🚧; to indicate "under
construction".  This is something that can be merged as it is not called
in production code.

Related to:

- https://github.com/scientist-softserv/adventist-dl/issues/430
- https://github.com/scientist-softserv/adventist-dl/issues/437

### CODE REVIEW

File: lib/bulkrax/rerun_errored_entries_for_importer_service.rb [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    def intialize(importer_id:, last_run_id: nil, error_classes: [], logger: Rails.logger)
+      @logger = logger
+      @importer = Bulkrax::Importer.find(importer_id)
+      if last_run_id
+        @last_run = @importer.importer_runs.find(last_run_id)
+      else
+        @last_run = @importer.last_run
+      end
+      @new_run = create_new_run_from_last_run
+      @error_classes = Array.wrap(error_class).map(&:to_s)
```
Comment (ShanaLMoore) [inline lib/bulkrax/rerun_errored_entries_for_importer_service.rb:35]: I have this running locally. I'm getting an undefined method error_class. error_classes => [ ] What should the error_classes be?

Comment (ShanaLMoore) [reply lib/bulkrax/rerun_errored_entries_for_importer_service.rb:35]: ah, I should be passes in the error class when calling for this service. Is there documentation for the options?

Comment (ShanaLMoore) [inline lib/bulkrax/rerun_errored_entries_for_importer_service.rb:35]:
```suggestion
      @error_classes = Array.wrap(error_classes).map(&:to_s)
```

should error class be error_classes ?

**Comment by ShanaLMoore** [comment]
I think something may be off. Locally I had a good example with failures. I wish I had taken a screenshot. When I called the service all of the records look to be pending now when I only expected the errored ones. 

image

File: lib/bulkrax/rerun_errored_entries_for_importer_service.rb [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    #   - When entry is an error, "enqueue" and create pending status
+    #   - When entry is success, "copy status to new run"
+    def call
+      start_processing_importer!
+
+      # Start processing the importer's entries.
+      #
+      # HACK: The 'Bulkrax::Entry' may be a bit presumptive so caller beware.
+      last_run.statuses.where(statusable_type: 'Bulkrax::Entry').find_each do |status|
+        process_status!(status: status)
```
Comment (ShanaLMoore) [inline lib/bulkrax/rerun_errored_entries_for_importer_service.rb:59]:
```suggestion
        processes_status!(status: status)
```

File: lib/bulkrax/rerun_errored_entries_for_importer_service.rb [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    private
+
+    ##
+    # This method should emulate the logic we perform when we begin processing an import.
+    def start_processing_importer!
+      # Reading through Bulkrax, I'm not seeing anything that needs doing.
+    end
+
+    def processes_status!(status:)
+      if status.error_classes.present?
```
Comment (ShanaLMoore) [inline lib/bulkrax/rerun_errored_entries_for_importer_service.rb:74]:
```suggestion
      if status.error_class.present?
```

**Review by ShanaLMoore — APPROVED** [review]
Tested this locally. It did not impact the other entries' statuses.

**Comment by jillpe** [comment]
closed - separate ticket created for clean up tasks

**Pull request merged**